### PR TITLE
Rewrite sizes of certain constant-sized arrays

### LINKED ISSUE #263: dredd apply mutation to expression used in variable initialisation with `const` qualifier
Applying dredd to the following file
```
void a() {
  const int b = 2;
  int c[b]{};
}
```
produce
```
void a() {
  const int b = __dredd_replace_expr_int_constant(42, 0);
  int c[b]{};
}
```

This result in compilation error:
```
variable-sized object may not be initialized
```

### PR DESCRIPTION
**Pull request by afd**
Avoids a problem where constant declarations that appear in the size expression for a constant-sized array might have their initialisers mutated, such that the size expression is no longer a compile time constant. This is a problem if the array is also initialized. The solution is to rewrite the size expressions of such arrays, replacing the original compile-time-constant expression with the value to which it evaluates. The constants that make up the expression - which may also be referred to elsewhere - can then have their initial values freely mutated.

Fixes #263.

### CODE REVIEW

File: src/libdredd/src/mutate_ast_consumer.cc [hunk 1]
```diff
@@ -90,6 +96,37 @@ void MutateAstConsumer::HandleTranslationUnit(clang::ASTContext& ast_context) {
...
+  // Rewrite the size expressions of constant-sized arrays as needed.
+  for (const auto& constant_sized_array_decl :
+       visitor_->GetConstantSizedArraysToRewrite()) {
+    assert(compiler_instance_->getLangOpts().CPlusPlus &&
+           constant_sized_array_decl->getType()->isConstantArrayType());
+    auto constant_array_typeloc = constant_sized_array_decl->getTypeSourceInfo()
+                                      ->getTypeLoc()
+                                      .getAs<clang::ConstantArrayTypeLoc>();
+    if (constant_array_typeloc.isNull()) {
+      // In some cases a declaration with constant array type does not have
```
Comment (JonathanFoo0523) [inline src/libdredd/src/mutate_ast_consumer.cc:108]: I don't fully get this. Can you provide an example?

File: src/libdredd/src/mutate_visitor.cc [hunk 2]
```diff
@@ -583,6 +583,19 @@ bool MutateVisitor::TraverseCompoundStmt(clang::CompoundStmt* compound_stmt) {
...
   var_decl_source_locations_.insert(var_decl->getLocation());
+
+  if (compiler_instance_->getLangOpts().CPlusPlus &&
+      var_decl->getType()->isConstantArrayType()) {
+    // We have a constant-sized C++ array. Some C++ compilers complain if
+    // components of the size expression are not compile-time constants. Other
+    // compilers are OK with this unless the array is also initialized. To allow
+    // the constant declarations that occur in a size expression to have their
+    // initial values mutated (because they may be used elsewhere), but to avoid
+    // compilation errors, we record the arrays in question to that later their
```
[the remark this answers is not in the document]
Comment (afd) [reply src/libdredd/src/mutate_visitor.cc:594]: Yes, fixed - thanks.

File: src/libdredd/src/mutate_visitor.cc [hunk 3]
```diff
@@ -583,6 +583,19 @@ bool MutateVisitor::TraverseCompoundStmt(clang::CompoundStmt* compound_stmt) {
...
+    // size expresisons can be replaced with the value to which the size
```
[the remark this answers is not in the document]
Comment (afd) [reply src/libdredd/src/mutate_visitor.cc:595]: Fixed, thanks.

File: test/execute/constant_sized_array/mutants.txt [hunk 4]
```diff
@@ -0,0 +1,3 @@
+0
```
[the remark this answers is not in the document]
Comment (afd) [reply test/execute/constant_sized_array/mutants.txt:1]: Good idea - done.

File: src/libdredd/src/mutate_ast_consumer.cc (hunk 1 above)
Comment (afd) [reply src/libdredd/src/mutate_ast_consumer.cc:108]: Rephrased to:

```
      // In some cases a declaration with constant array type does not have
      // an associated ConstantArrayTypeLoc object. This happens, for example,
      // when structured bindings are used. For example, consider:
      //
      //     auto [x, y] = a;
      //
      // This yields a constant-sized array, but there is no explicit array type
      // declaration. In such cases, no action is required.
```

**Pull request merged**